pkg/scheduler/framework/frameworkutils.go

```diff
@@ -0,0 +1,372 @@
+	dangling = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+	deleted = make([]*fleetv1beta1.ClusterResourceBinding, 0, len(bindings))
+
+	// Build a map for clusters for quick loopup.
+	clusterMap := make(map[string]fleetv1beta1.MemberCluster)
+	for _, cluster := range clusters {
+		clusterMap[cluster.Name] = cluster
+	}
+
+	for idx := range bindings {
+		binding := bindings[idx]
+		targetCluster, isTargetClusterPresent := clusterMap[binding.Spec.TargetCluster]
+
+		switch {
+		case binding.DeletionTimestamp != nil:
+			// Check if the binding has been deleted, but still has the scheduler cleanup finalizer in presence.
+			if controllerutil.ContainsFinalizer(&binding, fleetv1beta1.SchedulerCleanupFinalizer) {
+				deleted = append(deleted, &binding)
+			}
+			// Ignore any binding that has been deleted, and has no scheduler cleanup finalizer in presence.
+		case binding.Spec.State == fleetv1beta1.BindingStateDeleting:
+			// Ignore any binding that is of the deleting state.
+		case !isTargetClusterPresent || targetCluster.Spec.State == fleetv1beta1.ClusterStateLeave:
+			// Check if the binding is now obsolete, i.e., it is associated with a cluster that is no longer
+			// in normal operations, but is still of an active or creating state.
+			//
+			// Note that this check is solely for the purpose of detecting a situation where bindings are stranded
+			// on a leaving/left cluster; it does not perform any binding association eligibility check for the cluster.
+			dangling = append(dangling, &binding)
+		case binding.Spec.PolicySnapshotName != policy.Name:
+			// The binding is in the creating or active state, but is no longer associated with the latest scheduling policy snapshot.
+			obsolete = append(obsolete, &binding)
+		case binding.Spec.State == fleetv1beta1.BindingStateCreating:
+			// Check if the binding is of the creating state.
+			creating = append(creating, &binding)
+		case binding.Spec.State == fleetv1beta1.BindingStateActive:
+			// Check if the binding is of the active state.
+			active = append(active, &binding)
+		}
+	}
+
+	return active, creating, obsolete, dangling, deleted
+}
+
+// shouldDownscale checks if the scheduler needs to perform some downscaling, and (if so) how many active or creating bindings
+// it should remove.
+func shouldDownscale(policy *fleetv1beta1.ClusterPolicySnapshot, desired, present, obsolete int) (act bool, count int) {
+	if policy.Spec.Policy.PlacementType == fleetv1beta1.PickNPlacementType && desired <= present {
+		// Downscale only applies to CRPs of the Pick N placement type; and it only applies when the number of
+		// clusters requested by the user is less than the number of currently active + creating bindings combined;
+		// or there are the right number of active + creating bindings, yet some obsolete bindings still linger
+		// in the system.
+		if count := present - desired + obsolete; count > 0 {
```

Comment:
I don't fully get why we need to take account of obsolete as most bindings are obsolete by the time we do schedule (new policy)

Reply:
Hi Ryan! I mentioned the reasons in one of the above comments, PTAL. Essentially we have to do this to fix any incomplete scheduling cycles and to make sure that if a cluster is picked again no interruption will occur.